go/oasis-node/cmd/debug/txsource/workload/oversized.go

```diff
@@ -0,0 +1,100 @@
+package workload
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"google.golang.org/grpc"
+
+	"github.com/oasislabs/oasis-core/go/common/crypto/signature"
+	memorySigner "github.com/oasislabs/oasis-core/go/common/crypto/signature/signers/memory"
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	consensus "github.com/oasislabs/oasis-core/go/consensus/api"
+	"github.com/oasislabs/oasis-core/go/consensus/api/transaction"
+	runtimeClient "github.com/oasislabs/oasis-core/go/runtime/client/api"
+	staking "github.com/oasislabs/oasis-core/go/staking/api"
+)
+
+const (
+	NameOversized = "oversized"
+)
+
+var oversizedLogger = logging.GetLogger("cmd/txsource/workload/oversized")
+
+type oversized struct{}
+
+func (oversized) Run(gracefulExit context.Context, rng *rand.Rand, conn *grpc.ClientConn, cnsc consensus.ClientBackend, rtc runtimeClient.RuntimeClient) error {
+	txSignerFactory := memorySigner.NewFactory()
+	txSigner, err := txSignerFactory.Generate(signature.SignerEntity, rng)
+	if err != nil {
+		return fmt.Errorf("failed to generate signer key: %w", err)
+	}
+
+	ctx := context.Background()
+
+	// Fetch genesis consensus parameters.
+	// TODO: Don't dump everything, instead add a method to query just the parameters.
+	genesisDoc, err := cnsc.StateToGenesis(ctx, 1)
+	if err != nil {
+		return fmt.Errorf("failed to query state at genesis: %w", err)
+	}
+
+	var nonce uint64
+	fee := transaction.Fee{
+		Gas: 1000,
+	}
+	for {
+		// Generate a big transfer transaction which is valid, but oversized.
+		type customTransfer struct {
+			To   signature.PublicKey `json:"xfer_to"`
+			Data []byte              `json:"xfer_data"`
+		}
+		xfer := customTransfer{
+			// Send zero tokens to self, so the transaction will be valid.
+			To: txSigner.Public(),
+			// Include some extra random data so we are over the MaxTxSize limit.
+			Data: make([]byte, genesisDoc.Consensus.Parameters.MaxTxSize),
+		}
+		if _, err = rng.Read(xfer.Data); err != nil {
+			return fmt.Errorf("failed to generate bogus transaction: %w", err)
+		}
+
+		tx := transaction.NewTransaction(nonce, &fee, staking.MethodTransfer, &xfer)
+		signedTx, err := transaction.Sign(txSigner, tx)
+		if err != nil {
+			return fmt.Errorf("transaction.Sign: %w", err)
+		}
+		oversizedLogger.Debug("submitting oversized transaction",
+			"payload_size", len(xfer.Data),
+		)
+
+		// Wait for a maximum of 5 seconds as submission may block forever in case the client node
+		// is skipping all CheckTx checks.
+		submitCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		err = cnsc.SubmitTx(submitCtx, signedTx)
+		switch err {
+		case nil:
+			// This should never happen.
+			cancel()
+			return fmt.Errorf("successfully submitted an oversized transaction")
+		case consensus.ErrOversizedTx:
+			// Submitting an oversized transaction is an error, so we expect this to fail.
+			oversizedLogger.Info("transaction rejected due to ErrOversizedTx")
+		default:
```

Comment:
Could we ensure here that indeed timeout happened (and not some other unmatched error), probably another case matching on `context.Canceled`?

Reply:
The problem is that gRPC wraps this in its own errors so it's not actually `context.Canceled`. In any case, we shouldn't fail here in case of errors (the only failure is if the transaction is accepted), so the result would be the same.

Reply:
failing on other errors would ensure that this actually testing oversised tx's (not any other potential errors). However i do see that the `mux.maxTxSize` check is one of the first checks that we do, so it should probably be fine :+1: